From fc78c2134783ad0ae1cfb6b9434fe46c9e324871 Mon Sep 17 00:00:00 2001
From: Andy Polyakov <appro@openssl.org>
Date: Wed, 18 Jan 2017 23:17:30 +0000
Subject: [PATCH 6/6] crypto/evp: harden RC4_MD5 cipher.
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Originally a crash in 32-bit build was reported CHACHA20-POLY1305
cipher. The crash is triggered by truncated packet and is result
of excessive hashing to the edge of accessible memory (or bogus
MAC value is produced if x86 MD5 assembly module is involved). Since
hash operation is read-only it is not considered to be exploitable
beyond a DoS condition.

Thanks to Robert Święcki for report.

CVE-2017-3731

Change-Id: Ibc95a83f46ace9e1b58673f3217dc0f103e6120b
Reviewed-by: Rich Salz <rsalz@openssl.org>
---
 openssl/crypto/evp/e_rc4_hmac_md5.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/openssl/crypto/evp/e_rc4_hmac_md5.c b/openssl/crypto/evp/e_rc4_hmac_md5.c
index 5656319..dfae498 100644
--- a/openssl/crypto/evp/e_rc4_hmac_md5.c
+++ b/openssl/crypto/evp/e_rc4_hmac_md5.c
@@ -257,6 +257,8 @@ static int rc4_hmac_md5_ctrl(EVP_CIPHER_CTX *ctx, int type, int arg, void *ptr)
 
 		if (!ctx->encrypt)
 			{
+			if (len < MD5_DIGEST_LENGTH)
+				return -1;
 			len -= MD5_DIGEST_LENGTH;
 			p[arg-2] = len>>8;
 			p[arg-1] = len;
-- 
2.7.4

